service/controller/resource/ipam/azure_machinepool_checker.go

```diff
@@ -0,0 +1,80 @@
+package ipam
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	capiv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type AzureMachinePoolCheckerConfig struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+type AzureMachinePoolChecker struct {
+	ctrlClient client.Client
+	logger     micrologger.Logger
+}
+
+func NewAzureMachinePoolChecker(config AzureMachinePoolCheckerConfig) (*AzureMachinePoolChecker, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AzureMachinePoolChecker{
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+	}
+
+	return a, nil
+}
+
+func (c *AzureMachinePoolChecker) Check(ctx context.Context, namespace string, name string) (bool, error) {
+	azureMachinePool := &v1alpha3.AzureMachinePool{}
+	err := c.ctrlClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, azureMachinePool)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, c.ctrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	azureCluster, err := c.getAzureClusterFromCluster(ctx, cluster)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	// In case there is no subnet tracked so far, we want to proceed with the allocation process.
+	for _, subnet := range azureCluster.Spec.NetworkSpec.Subnets {
+		if subnet.Name == azureMachinePool.Name {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (c *AzureMachinePoolChecker) getAzureClusterFromCluster(ctx context.Context, cluster *capiv1alpha3.Cluster) (*capzv1alpha3.AzureCluster, error) {
```

Comment:
We have this function repeated in 4 places in the project, I'll place it in some common package so it's reused.

Reply:
I think there was something towards the owner reference functionality. Maybe check the spark PR. If not, maybe all of those could be refactored to same?

Reply:
Thanks, I'll check Spark CR then first.

Reply:
There is a function in Spark CR for getting a owner `Cluster`, but here we need to get `AzureCluster` for the owner `Cluster`, so the other way around.